coordinator/src/referrals.rs

```diff
@@ -0,0 +1,436 @@
+) -> Result<ReferralStatus> {
+    let bonus_status = db::bonus_status::active_status_for_user(connection, &trader_pubkey)?;
+    let user = db::user::get_user(connection, &trader_pubkey)?.context("User not found")?;
+
+    // first we check for promo, promo always has highest priority
+    if let Some(status) = has_active_promo_bonus(connection, &bonus_status, &user)? {
+        return Ok(status);
+    }
+
+    // then we check if he was referred, if he has been, his referral bonus might still be active
+    // first we check for promo, promo always has highest priority
+    if let Some(status) = has_active_referent_bonus(connection, &bonus_status, &user)? {
+        return Ok(status);
+    }
+
+    // if none of the above, we check for the highest tier without promo or referent
+    let mut filtered_boni = bonus_status
+        .iter()
+        .filter(|bonus| {
+            bonus.bonus_type != BonusType::Promotion && bonus.bonus_type != BonusType::Referent
+        })
+        .collect::<Vec<_>>();
+
+    // we sort by tier_level, higher tier means better bonus
+    filtered_boni.sort_by(|a, b| b.tier_level.cmp(&a.tier_level));
+
+    // next we pick the highest
+    if let Some(bonus) = filtered_boni.first() {
+        let referrals =
+            db::bonus_tiers::all_referrals_by_referring_user(connection, &trader_pubkey)?;
+        let tier = db::bonus_tiers::tier_by_tier_level(connection, bonus.tier_level)?;
+
+        return Ok(ReferralStatus {
+            referral_code: user.referral_code,
+            number_of_activated_referrals: referrals
+                .iter()
+                .filter(|referral| {
+                    referral.referred_user_total_quantity.floor() as i32
+                        > tier.min_volume_per_referral
+                })
+                .count(),
+            number_of_total_referrals: referrals.len(),
+            referral_tier: tier.tier_level as usize,
+            referral_fee_bonus: Decimal::from_f32(tier.fee_rebate).expect("to fit"),
+            bonus_status_type: Some(bonus.bonus_type.into()),
+        });
+    }
+
+    // None of the above, user is a boring normal user
+    Ok(ReferralStatus::new(trader_pubkey))
+}
+
+fn has_active_referent_bonus(
+    connection: &mut PooledConnection<ConnectionManager<PgConnection>>,
+    bonus_status: &Vec<BonusStatus>,
+    user: &User,
+) -> Result<Option<ReferralStatus>> {
+    if let Some(bonus) = bonus_status
+        .iter()
+        .find(|status| BonusType::Referent == status.bonus_type)
+    {
+        tracing::debug!(
+            trader_pubkey = bonus.trader_pubkey,
+            bonus_type = ?bonus.bonus_type,
+            tier_level = bonus.tier_level,
+            "Trader was referred");
+        let tier = db::bonus_tiers::tier_by_tier_level(connection, bonus.tier_level)?;
+
+        return Ok(Some(ReferralStatus {
+            referral_code: user.referral_code.clone(),
+            number_of_activated_referrals: 0,
```

Comment:
❓ what if the user referred users and is part of a promo. In that case that value wouldn't be correct.

Reply:
I've removed the promo